tests/model/ChangeSetTest.php

```diff
@@ -0,0 +1,439 @@
+<?php
+
+/**
+ * Provides a set of targettable permissions for tested models
+ *
+ * @mixin Versioned
+ * @mixin DataObject
+ */
+trait ChangeSetTest_Permissions {
+	public function canEdit($member = null) {
+		return $this->can(__FUNCTION__, $member);
+	}
+
+	public function canDelete($member = null) {
+		return $this->can(__FUNCTION__, $member);
+	}
+
+	public function canCreate($member = null, $context = array()) {
+		return $this->can(__FUNCTION__, $member, $context);
+	}
+
+	public function canPublish($member = null, $context = array()) {
+		return $this->can(__FUNCTION__, $member, $context);
+	}
+
+	public function canUnpublish($member = null, $context = array()) {
+		return $this->can(__FUNCTION__, $member, $context);
+	}
+
+	public function can($perm, $member = null, $context = array()) {
+		$perms = [
+			"PERM_{$perm}",
+			'CAN_ALL',
+		];
+		return Permission::checkMember($member, $perms);
+	}
+}
+
+/**
+ * @mixin Versioned
+ */
+class ChangeSetTest_Base extends DataObject implements TestOnly {
+	use ChangeSetTest_Permissions;
+
+	private static $db = [
+		'Foo' => 'Int',
+	];
+
+	private static $has_many = [
+		'Mids' => 'ChangeSetTest_Mid',
+	];
+
+	private static $owns = [
+		'Mids',
+	];
+
+	private static $extensions = [
+		"Versioned",
+	];
+}
+
+/**
+ * @mixin Versioned
+ */
+class ChangeSetTest_Mid extends DataObject implements TestOnly {
+	use ChangeSetTest_Permissions;
+
+	private static $db = [
+		'Bar' => 'Int',
+	];
+
+	private static $has_one = [
+		'Base' => 'ChangeSetTest_Base',
+		'End' => 'ChangeSetTest_End',
+	];
+
+	private static $owns = [
+		'End',
+	];
+
+	private static $extensions = [
+		"Versioned",
+	];
+}
+
+/**
+ * @mixin Versioned
+ */
+class ChangeSetTest_End extends DataObject implements TestOnly {
+	use ChangeSetTest_Permissions;
+
+	private static $db = [
+		'Baz' => 'Int',
+	];
+
+	private static $extensions = [
+		"Versioned",
+	];
+}
+
+/**
+ * Test {@see ChangeSet} and {@see ChangeSetItem} models
+ */
+class ChangeSetTest extends SapphireTest {
+
+	protected static $fixture_file = 'ChangeSetTest.yml';
+
+	protected $extraDataObjects = [
+		'ChangeSetTest_Base',
+		'ChangeSetTest_Mid',
+		'ChangeSetTest_End',
+	];
+
+	/**
+	 * Automatically publish all objects
+	 */
+	protected function publishAllFixtures() {
+		$this->logInWithPermission('ADMIN');
+		foreach($this->fixtureFactory->getFixtures() as $class => $fixtures) {
+			foreach ($fixtures as $handle => $id) {
+				$this->objFromFixture($class, $handle)->doPublish();
```

Comment:
The alternative I considered was to add a new parameter to doPublish($recursive) to flag whether or not related objects should be published. This would be left true (default) when doing single page publishing, but changesets could set it to false, since it handles related object publishing itself.

Would you prefer this api change to the one we're using?